[SPARK-3957]: show broadcast variable resource usage info in UI

CodingCat opened the pull request:
show the broadcast variables resource usage in UI

![image](https://cloud.githubusercontent.com/assets/678008/4696076/6bb23818-57e5-11e4-83b9-a59ced0d1bc4.png)

Commit:
update page

CodingCat committed:
report broadcast var in UI

CodingCat committed:
update page

Author:   CodingCat  <zhunansjtu@gmail.com>

CodingCat committed:
send broadcastInfo in blockManager heartbeat

Commit:
send broadcastInfo in blockManager heartbeat

CodingCat committed:
including broadcast variable mem usage in executor page

CodingCat committed:
separate broadcast var from nonRddBlocks

CodingCat committed:
storagePage

CodingCat committed:
code clean

CodingCat committed:
fix compile error

CodingCat committed:
fix compile error

CodingCat committed:
remove Heartbeat piggyback mechanism and add async UpdateBlockInfo

CodingCat committed:
add broadcast info

CodingCat committed:
fix compile issue

CodingCat committed:
code clean

CodingCat committed:
scala style fix

CodingCat committed:
fix test case

CodingCat committed:
fix test cases

CodingCat committed:
fix livelistenerBus issue

CodingCat committed:
make mima happy

CodingCat committed:
remove unnecessary imports

SparkQA commented:
[Test build #25600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25600/consoleFull) for   PR 2851 at commit [`bf7095c`](https://github.com/apache/spark/commit/bf7095cc31d4434daabba3bfcfecec98c9389dbb).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25600/consoleFull) for   PR 2851 at commit [`bf7095c`](https://github.com/apache/spark/commit/bf7095cc31d4434daabba3bfcfecec98c9389dbb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25600/
Test PASSed.

CodingCat commented:
ping

SparkQA commented:
[Test build #25817 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25817/consoleFull) for   PR 2851 at commit [`1f53135`](https://github.com/apache/spark/commit/1f5313515f6151ee4d8d24cefb4ae1f870703f7e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25817 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25817/consoleFull) for   PR 2851 at commit [`1f53135`](https://github.com/apache/spark/commit/1f5313515f6151ee4d8d24cefb4ae1f870703f7e).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25817/
Test FAILed.

CodingCat committed:
fix compile issue

SparkQA commented:
[Test build #25837 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25837/consoleFull) for   PR 2851 at commit [`6b25507`](https://github.com/apache/spark/commit/6b255078375b1861fd252e3db365c73d38ac1945).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25837 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25837/consoleFull) for   PR 2851 at commit [`6b25507`](https://github.com/apache/spark/commit/6b255078375b1861fd252e3db365c73d38ac1945).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25837/
Test PASSed.

andrewor14 commented:
Hey @CodingCat this seems like a very complicated change for showing broadcast variables. In particular I do not understand why we need to introduce a new listener event for this. Can you summarize how the broadcast blocks info is sent to the driver before and after this PR?

SparkQA commented:
[Test build #27377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27377/consoleFull) for   PR 2851 at commit [`3a0aa25`](https://github.com/apache/spark/commit/3a0aa258a430217cc2dc69b11531230d078b16a7).
 * This patch merges cleanly.

CodingCat commented:
Hey, @andrewor14 

Sorry for the late response, I just cleaned a congested task queue in my schedule...I will address your comments within today

SparkQA commented:
[Test build #27377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27377/consoleFull) for   PR 2851 at commit [`3a0aa25`](https://github.com/apache/spark/commit/3a0aa258a430217cc2dc69b11531230d078b16a7).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27377/
Test FAILed.

CodingCat committed:
fix compile issue

SparkQA commented:
[Test build #27380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27380/consoleFull) for   PR 2851 at commit [`b65d49e`](https://github.com/apache/spark/commit/b65d49e1fa4e00e7353851197b9fce92a3e911e1).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27380 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27380/consoleFull) for   PR 2851 at commit [`b65d49e`](https://github.com/apache/spark/commit/b65d49e1fa4e00e7353851197b9fce92a3e911e1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27380/
Test FAILed.

CodingCat committed:
compilation issue

SparkQA commented:
[Test build #27387 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27387/consoleFull) for   PR 2851 at commit [`bf03271`](https://github.com/apache/spark/commit/bf032716ed7f7d8d57e44c936e70385ab775c6c8).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27387 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27387/consoleFull) for   PR 2851 at commit [`bf03271`](https://github.com/apache/spark/commit/bf032716ed7f7d8d57e44c936e70385ab775c6c8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27387/
Test FAILed.

CodingCat committed:
fix the bug when updating non-broadcast variable

SparkQA commented:
[Test build #27404 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27404/consoleFull) for   PR 2851 at commit [`5c60a34`](https://github.com/apache/spark/commit/5c60a345e9b6e34f51403df1bf721d00fac7bffa).
 * This patch merges cleanly.

CodingCat committed:
remove unnecessary variable

SparkQA commented:
[Test build #27405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27405/consoleFull) for   PR 2851 at commit [`08ad98d`](https://github.com/apache/spark/commit/08ad98d5a544225c733411fb895de5fa32431c24).
 * This patch merges cleanly.

CodingCat commented:
Hi, @andrewor14, thanks for the review.

To answer your question:

> how the broadcast blocks info is sent to the driver before and after this PR?

before this PR, the broadcast blocks was not sent to the driver from executors, 

in this PR, I am exactly using the strategy mentione by @shivaram, i.e. setting tellMaster=true

> Would all the new logic just go away if we simply set tellMaster = true for all broadcast blocks,

No...the complexity comes from 1) updating the page, and 2) as mentioned by @rxin , we need to avoid introduce more complexity to BlockManagers (at the beginning I simply piggyback the broadcast info in the BlockManager heartbeat)

to address 2), I use the listenerBus to notify the StorageStatusListener and StorageListener that some of the blocks are updated, they need to apply the operation respectively (StorageStatusListener => update executorIdToStorageStatus, StorageListener => update internal data structure)

SparkQA commented:
[Test build #27404 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27404/consoleFull) for   PR 2851 at commit [`5c60a34`](https://github.com/apache/spark/commit/5c60a345e9b6e34f51403df1bf721d00fac7bffa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27404/
Test FAILed.

SparkQA commented:
[Test build #27405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27405/consoleFull) for   PR 2851 at commit [`08ad98d`](https://github.com/apache/spark/commit/08ad98d5a544225c733411fb895de5fa32431c24).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27405/
Test FAILed.

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -81,19 +84,38 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
+   *
+   * Note that this is somewhat expensive, as it involves cloning the underlying maps and then
+   * concatenating them together. Much faster alternatives exist for common operations such as
+   * getting the memory, disk, and off-heap memory sizes occupied by this RDD.
```
should be "broadcast variable" instead of "RDD"

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -118,8 +140,20 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
+        // Actually remove the block, if it exists
+        if (_broadcastBlocks.contains(broadcastId)) {
+          val removed = _broadcastBlocks(broadcastId).remove(blockId)
+          // If the given RDD has no more blocks left, remove the RDD
```
again, broadcast variable instead of RDD

squito replied:
actually now I am thinking there is a lot of copy-pasting that could be cleaned up.  you could mostly merge this w/ the block above if you did something like:

```
case rddOrBlockId @ (_: BroadcastBlockId | _: RDDBlockId) =>
  val (id, blockMap) = getIdAndBlockMap(rddOrBlockId)
  val removed = blockMap(id).remove(blockId)
  ...
```

where the helper function `getIdAndBlockMap` is something like:

```
def getIdAndBlockMap(blockId: BlockId): (Int, Map[BlockId, BlockStatus]) = blockId match {
  case RDDBlockId(rddId, _) => (rddId, _rddBlocks)
  case BroadcastBlockId(broadcastId, _) => (broadcastId, _broadcastBlocks)
}
```

and then you could do a similar thing in a few other places.  You could also take this a step further, and even merge `_rddBlocks` and `_broadcastBlocks` into a `EnumMap[BlockType, Map[BlockId, BlockStatus]]` if you made a new `public enum BlockType{ RDD,Broadcast}`, but that might not really help much since at the end of the day you do want separate getter methods for the RDD and Broadcast stuff

squito replied:
you could tighten it up even more if you moved that helper method to an extractor:

```
object RddOrBlockId {
  def unapply(blockId: BlockId): Option[(Int, Map[BlockId,BlockStatus])] = {
    case RDDBlockId(rddId, _) => Some((rddId, _rddBlocks))
    case BroadcastBlockId(broadcastId, _) => Some((broadcastId, _broadcastBlocks))
    case _ => None
  }
}
```

then the actual matching in each of these blocks would simplify to:

```
  case RddOrBlockId(id, blockMap) =>
    val removed = blockMap(id).remove(blockId)
    ...
```
(the point is, the helper function to get the id & blockMap goes right into the pattern match.)  But maybe this is a bit of scala that is too esoteric for our code base for saving one line in each of these cases.

rxin replied:
Please, don't do that. It is very hard to understand these kinds of extractors, especially when RddOrBlockId itself doesn't have a real class.

CodingCat commented:
Hi, @squito , thanks for the review, I will handle it in the weekend

squito commented on core/src/main/scala/org/apache/spark/ui/storage/BroadcastPage.scala:
```diff
@@ -0,0 +1,152 @@
+
+import scala.xml.Node
+
+private[ui] class BroadcastPage(parent: StorageTab) extends WebUIPage("broadcast"){
```
this is just a minor variation of `RDDPage`, right?  better to make a common superclass with most of the code, then `RDDPage` and `BroadcastPage` would just need a couple of tweaks (eg., labels in the UI, which block map to look at etc.)

squito commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -91,6 +91,7 @@ private[spark] object JsonProtocol {
         executorRemovedToJson(executorRemoved)
       // These aren't used, but keeps compiler happy
       case SparkListenerExecutorMetricsUpdate(_, _) => JNothing
+      case SparkListenerBlockUpdate(_, _, _) => JNothing
```
This event should be converted to json so it can be seen in the UI for completed apps.  You need to implement serialization here, and also deserialization in `sparkEventFromJson`, as that is used by `ReplayListenerBus` to show the UI for completed apps.

andrewor14 replied:
how often is this event posted, is it once per cached partition? If so there might be too many such events to log them.

squito commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterActor.scala:
```diff
@@ -466,10 +467,11 @@ private[spark] class BlockManagerInfo(
+      tachyonSize: Long): BlockStatus =  {
 
-    updateLastSeenMs()
+    var blockStatus: BlockStatus = null
```
we can get rid of this var -- just have each branch of the if/else finish w/ the block status you want to return. 

```
if (_blocks.containsKey(blockId)) {
  ...
  if (storageLevel.isValid) {
    ...
   _blocks.get(blockId)
  } else if (_blocks.containsKey(blockId)) {//I realize you aren't modifying this part, but isn't this condition redundant?
    val blockStatus: BlockStatus = _blocks.get(blockId)
    ...
    BlockStatus(storageLevel, 0, 0, 0)
  }
} else {
  null
}
```

squito commented on core/src/main/scala/org/apache/spark/storage/BroadcastInfo.scala:
```diff
@@ -0,0 +1,52 @@
+      if (this.id == that.id) {
+        0
+      }
+      -1
```
this can be just `this.id - that.id`

vanzin replied:
what type does `compare` return? It's generally an `int`, and `id` is a long. (BTW, *hint*hint*, make the method's return type explicit.)

squito commented:
Hi @CodingCat ,

thanks for working on this, sorry the review has been dragging out.  This will be a nice addition, but I have some concerns similar to the other reviewers -- but lemme explain in a little more detail.  I don't necessarily see anything wrong with your approach for getting the Broadcast block info by itself, but it seems like it ought to share a similar code path to tracking the RDDBlocks.  RDDBlocks get updated in the `SparkListenerTaskEnd` event, through `event.taskMetrics.updatedBlocks`.  this makes its way to the current `StorageTab` in [`StorageListener`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala#L59), which then [filters down to the RDDBlocks](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala#L50) via `BlockId.asRDDId`.

So, this makes me wonder about two possible alternatives for unifying the code paths:

1) Can you get the Broadcast block ids into the same `SparkListenerTaskEnd` event, which would eliminate the need to create a new event?  (In fact, maybe those block ids are already there?  I am still looking into the code for how those events get created ...)

2) If we can't get Broadcast block ids into the same `SparkListenerTaskEnd`, does it make sense to change the RDD block tracking logic to do the same thing?  Yes this will be a bigger change, but it seems strange to have these two blocks tracked by such different mechanisms so it will lead to simpler code overall.  (But really that pushes me to option #1)

squito commented:
hmm, ok now I see that it is hard to get the BroadcastBlockIds into the task end metrics, since when broadcast blocks get created, its a side effect of task-deserialization, so the broadcast var doesn't really know which task its being created for.  The only other option I can think of is to create a `ThreadLocal` with the current task id, before deserializing the task [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L177).  Then when the broadcast var is actually put in the block manager, ([eg.  for `TorrentBroadcast`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala#L181)) and finally when the task is finished you'd [update the metrics](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L214).

Honestly I am somewhat torn between (a) leaving it as you have it, (b) what I've outlined in this comment, and (c) having RDD block info also tracked by your new mechansim (option 2 above)

CodingCat commented:
working on this now

squito commented:
well I was leaning towards using a `ThreadLocal` to get the broadcast blocks into the task end event ... but I forgot that broadcast blocks are also created by the driver (eg. [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala#L98) or [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/broadcast/HttpBroadcast.scala#L52)). It doesn't make any sense to stick those into a task end event.  I suppose they could go into a task **start** event, but that is still a little weird, maybe we really do need a new event.

@rxin @andrewor14 thoughts?

CodingCat committed:
tighten the methods implementation in StorageUtils and RDD page & BroadcastPage

CodingCat committed:
get rid of mutable variable

SparkQA commented:
[Test build #27569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27569/consoleFull) for   PR 2851 at commit [`57d3f46`](https://github.com/apache/spark/commit/57d3f4609d069c8891cda3f0658d1eb1bb93e9f1).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27569/consoleFull) for   PR 2851 at commit [`57d3f46`](https://github.com/apache/spark/commit/57d3f4609d069c8891cda3f0658d1eb1bb93e9f1).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27569/
Test FAILed.

CodingCat committed:
add license info

SparkQA commented:
[Test build #27570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27570/consoleFull) for   PR 2851 at commit [`b05196a`](https://github.com/apache/spark/commit/b05196af57e2c9ad8504b7ab12f1d3b2ea454067).
 * This patch merges cleanly.

CodingCat committed:
apply event filter

SparkQA commented:
[Test build #27571 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27571/consoleFull) for   PR 2851 at commit [`34aed25`](https://github.com/apache/spark/commit/34aed25737ef896356505e4ca2f50dd2d788b98c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27570/consoleFull) for   PR 2851 at commit [`b05196a`](https://github.com/apache/spark/commit/b05196af57e2c9ad8504b7ab12f1d3b2ea454067).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27570/
Test FAILed.

CodingCat committed:
implement serialization and deserialization of BlockUpdateEvent

CodingCat committed:
addressed the failed test cases

SparkQA commented:
[Test build #27588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27588/consoleFull) for   PR 2851 at commit [`04726c4`](https://github.com/apache/spark/commit/04726c410909144581a308651837c2039b841276).
 * This patch merges cleanly.

CodingCat commented:
Hi, @squito, thank you very much for your patient review
 
 Here is the summary on what I have done on addressing your comments
 
 1. regarding the duplication in RDDPage and BroadcastPage
 
    Now, both of them inherit from InMemoryObjectPage, which defines the common parts of these two classes, e.g. Worker Header, etc.
 
 2. Serialization and Deserialization of SparkListenerBlockUpdate event
 
    I implemented the serialization and deserialization methods (https://github.com/apache/spark/pull/2851/files#diff-4f6ba18259eb4c31ac930e18f1ba6f88R99) and (https://github.com/apache/spark/pull/2851/files#diff-4f6ba18259eb4c31ac930e18f1ba6f88R480)

  	To prevent the large volume of events being logged, there is an event filter existing in SparkListenerBus (https://github.com/apache/spark/pull/2851/files#diff-01185688f339238fc6689baa6106df63R27). Now this filter ensure that only the broadcast block update event is logged (https://github.com/apache/spark/pull/2851/files#diff-fbe8f967070627c8dc155237e77c7314R130) (actually this was implemented long time ago)
  	
 3. get rid of the mutable variable in BlockManagerMasterActor 
 	
 	Done (https://github.com/apache/spark/pull/2851/files#diff-6da9360efd5e37dcd0edcc651db1e9cbR508)
 	
 4. shorten the compare() in BroadcastInfo (https://github.com/apache/spark/pull/2851/files#diff-6da9360efd5e37dcd0edcc651db1e9cbR508)
 
    Unfortunately, I probably cannot do that, since compare() returns Int, but broadcastId is Long
  	
 5. tighten the methods in StorageUtils
 
    Done (https://github.com/apache/spark/pull/2851/files#diff-6da9360efd5e37dcd0edcc651db1e9cbR508)

CodingCat commented:
Potential problems in the current patch

1. to de-duplicate the code in StorageUtils, I save RDD ID with **Int** type in a **Long** Field (https://github.com/apache/spark/pull/2851/files#diff-21027f5c826cd378daaae5f7c3eea2b5R41), otherwise it's hard to get a BlockMap according to the blockId (https://github.com/apache/spark/pull/2851/files#diff-21027f5c826cd378daaae5f7c3eea2b5R126) (you will see a lot of type imcompatibility from the compiler)

2. InMemoryObjectPage.scala (https://github.com/apache/spark/pull/2851/files#diff-dc63d4ab5b2ed5a2135c054be78820eeR97), I'm not sure if the semantics of (memory usage of a worker) is correct...it should be the memory used by the current InMemoryObject or the total memory usage?

SparkQA commented:
**[Test build #27571 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27571/consoleFull)**     for PR 2851 at commit [`34aed25`](https://github.com/apache/spark/commit/34aed25737ef896356505e4ca2f50dd2d788b98c)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27571/
Test FAILed.

CodingCat committed:
address failed test cases

SparkQA commented:
[Test build #27595 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27595/consoleFull) for   PR 2851 at commit [`0d854b8`](https://github.com/apache/spark/commit/0d854b835db225baa3b2ee2271d03afa87d9a3e8).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #27588 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27588/consoleFull)**     for PR 2851 at commit [`04726c4`](https://github.com/apache/spark/commit/04726c410909144581a308651837c2039b841276)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27588/
Test FAILed.

CodingCat commented:
weird...I can pass the test cases in my laptop without any problem....

CodingCat commented:
oh...the failed one is the previous submission.....

SparkQA commented:
[Test build #27595 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27595/consoleFull) for   PR 2851 at commit [`0d854b8`](https://github.com/apache/spark/commit/0d854b835db225baa3b2ee2271d03afa87d9a3e8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerBlockUpdate(blockManagerId: BlockManagerId, blockId: BlockId,`
  * `class DefaultSparkListenerEventFilter `
  * `class BroadcastInfo(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27595/
Test PASSed.

squito commented:
Hi @CodingCat 

thanks for making all the updates.  Sorry I hadn't realized the subtlety w/ `Int` vs `Long` on the `RDDBlockId` and `BroadcastBlockId`.  Still, I think its still a good change for simplifying the code, glad you figured a way to make it work.   And on issue (2), the memory usage of a worker, I think what you have is correct, its supposed to be memory usage of the particular object on the worker.  This is on a UI page in the context of a particular object -- there is a separate page to summarize the memory usage of the executor overall.  (though I agree the UI is a little confusing ...)

I'll make a few more minor comments on the code, but mostly I just want to get another opinion on the right events to pass the block added events around, as I have mentioned above.

squito commented on core/src/main/scala/org/apache/spark/scheduler/SparkListenerBus.scala:
```diff
@@ -24,7 +24,13 @@ import org.apache.spark.util.ListenerBus
+
   override def onPostEvent(listener: SparkListener, event: SparkListenerEvent): Unit = {
+    if (!filter.validate(event)) {
+      return  
```
I think the `DefaultSparkListenerEventFilter` is probably adding an abstraction without any really good reason.  If we do stick w/ the new `SparkListenerBlockUpdate` events, I think its better if you just put the filter into the right method on the event logging listener and remove `DefaultSparkListenerEventFilter`.

squito replied:
also, I think there is still one missing piece to get the json into the event logging for the history server.  You need to implement to put in the implementation for `onBlockUpdate` in `EventLoggingListener`.  I am suggesting that you just put this filter into that implementation, and get rid of the `Filter` abstraction.

squito commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterActor.scala:
```diff
@@ -522,7 +523,9 @@ private[spark] class BlockManagerInfo(
       }
+      return BlockStatus(storageLevel, 0, 0, 0)
     }
+    null
```
you don't need `return` here -- the last value of each block is its return value.  so this could be:

```
if (storageLevel.isValid) {
 ...
  _blocks.get(blockId)
} else if (_blocks.containsKey(blockId)) {
  ...
  BlockStatus(storageLevel, 0, 0, 0)
} else {
  null
}
```

squito replied:
actually, can you explain the `null` case?  how does that happen, and won't that result in an NPE when it gets to your code in `StorageStatusListener` and in `JsonProtocol`?

CodingCat replied:
actually the code should not enter into this line...unless something wrong happened in updating https://github.com/CodingCat/spark/blob/SPARK-3957/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterActor.scala#L457

maybe I can change https://github.com/CodingCat/spark/blob/SPARK-3957/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterActor.scala#L373 to an Option, or add a if guard in https://github.com/CodingCat/spark/blob/SPARK-3957/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterActor.scala#L373

squito commented on core/src/main/scala/org/apache/spark/storage/RDDInfo.scala:
```diff
@@ -21,13 +21,14 @@ import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.rdd.RDD
 import org.apache.spark.util.Utils
 
+trait InMemoryObjectInfo
```
you don't really need this trait at all

CodingCat replied:
Actually I used it in https://github.com/apache/spark/pull/2851/files#diff-dc63d4ab5b2ed5a2135c054be78820eeR47

I need a common base class to save RDD and Broadcast Info ...

squito commented on core/src/main/scala/org/apache/spark/storage/RDDInfo.scala:
```diff
@@ -49,9 +50,40 @@ class RDDInfo(
+      if (this.id == that.id) {
+        return 0
+      }
+      -1
```
super minor:
```
if (this.id > that.id) {
  1
}  else if (this.is == that.id) {
  0
}  else {
  -1
}
```
(sorry I was wrong w/ the earlier suggestion)

vanzin replied:
If you really want a suggestion here :-):

    com.google.common.primitives.Longs.compare(this.id, that.id)

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -271,4 +368,19 @@ private[spark] object StorageUtils {
 
+
+  /**
+   * Return a mapping from block ID to its locations for each block that belongs to the given RDD.
```
broadcast block, not RDD

squito commented on core/src/main/scala/org/apache/spark/ui/storage/InMemoryObjectPage.scala:
```diff
@@ -0,0 +1,123 @@
+
+import scala.xml.Node
+
+private[ui] abstract class InMemoryObjectPage(pageName: String, parent: StorageTab)
```
RDD aren't necessarily in memory ... maybe a better name would be `StorageDetailPage`?

squito commented on core/src/main/scala/org/apache/spark/ui/storage/BroadcastPage.scala:
```diff
@@ -0,0 +1,90 @@
+          <ul class="unstyled">
+            <li>
+              <strong>Total Partitions:</strong>
+              {objectInfo.numPartitions}
```
`numPartitions` doesn't mean anything for a Broadcast var.  I think its always `0` in your code, unless I've missed something.

squito commented:
can you also post a screenshot of the detailed page for a broadcast var?  Ideally involving a broadcast var that gets turned into multiple blocks by `TorrentBroadcast`, I think just needs to be over 4MB by default.

CodingCat commented:
![image](https://cloud.githubusercontent.com/assets/678008/6235037/622218b4-b6ad-11e4-8cd6-bd291fd4e58a.png)

CodingCat committed:
address the comments except filter one

CodingCat committed:
move filter logic to EventLoggingListener

SparkQA commented:
[Test build #27638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27638/consoleFull) for   PR 2851 at commit [`564f7d2`](https://github.com/apache/spark/commit/564f7d26de6337bfd1fb6453c6344d73c9dd0ed0).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27638/consoleFull) for   PR 2851 at commit [`564f7d2`](https://github.com/apache/spark/commit/564f7d26de6337bfd1fb6453c6344d73c9dd0ed0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27638/
Test PASSed.

CodingCat commented:
Hi, @squito , thanks for the further comments

I have addressed them and also replied under some of them

The snapshots was uploaded, 

I will not be here tomorrow but will be back after that....

CodingCat committed:
fix rebase mistake

SparkQA commented:
[Test build #28529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28529/consoleFull) for   PR 2851 at commit [`2ea8ee3`](https://github.com/apache/spark/commit/2ea8ee341a40ae543284875b3e6ba333fc07ca9d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28529/consoleFull) for   PR 2851 at commit [`2ea8ee3`](https://github.com/apache/spark/commit/2ea8ee341a40ae543284875b3e6ba333fc07ca9d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28529/
Test FAILed.

CodingCat committed:
fix rebase mistake

SparkQA commented:
[Test build #28534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28534/consoleFull) for   PR 2851 at commit [`223cdfb`](https://github.com/apache/spark/commit/223cdfb6c02412f6bafbbbe32e160f08a96e984f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28534/consoleFull) for   PR 2851 at commit [`223cdfb`](https://github.com/apache/spark/commit/223cdfb6c02412f6bafbbbe32e160f08a96e984f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28534/
Test PASSed.

CodingCat commented:
Hi, @squito , would you mind taking further review when you have time?

squito commented on core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:
```diff
@@ -67,10 +67,8 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, scheduler: TaskSchedule
-      val response = HeartbeatResponse(reregisterBlockManager = unknownExecutor)
-      executorLastSeen(executorId) = System.currentTimeMillis()
+      val response = HeartbeatResponse(
+        !scheduler.executorHeartbeatReceived(executorId, taskMetrics, blockManagerId))
```
did you mean to remove the update of `executorLastSeen`?  seems like maybe a mistake in the merge?

squito commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -181,6 +181,12 @@ private[spark] class EventLoggingListener(
   override def onExecutorRemoved(event: SparkListenerExecutorRemoved) =
     logEvent(event, flushLogger = true)
+  override def onBlockUpdate(event: SparkListenerBlockUpdate) = {
+    // we only log Broadcast block update for now
```
can you add to this comment that we don't need to log RDD blocks, b/c they are already logged as part of `StageCompleted` events

squito replied:
er, I guess its also happening in `TaskEnd` events

squito commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -357,7 +357,13 @@ private[spark] class BlockManager(
+      blockId.isBroadcast
+    }
+    // asynchronously report broadcast variables, in future we only need to change the
+    // implementation of ifAsync
```
I find this comment a little bit confusing -- it sounds like you are saying there is future work still to do.  Do you mean something like "Broadcast variables are currently reported asynchronously, but if we later decide they should be reported synchronously, we just need to update ifAsync`"

CodingCat replied:
hmmm...I mean, for reporting broadcast, we do it asynchronously, but in future, if we want to report other blocks, we can make ifAsync logic more complicate, e.g.

```

if (blockId.isBroadcast) true
else false

```

squito commented on core/src/main/scala/org/apache/spark/storage/StorageStatusListener.scala:
```diff
@@ -88,4 +88,19 @@ class StorageStatusListener extends SparkListener {
 
+  override def onBlockUpdate(blockUpdateEvent: SparkListenerBlockUpdate) = synchronized {
+    val executorId = blockUpdateEvent.blockManagerId.executorId
+    updateStorageStatus(executorId, Seq((blockUpdateEvent.blockId,
```
If I'm following things correctly, this will also get called for RDD blocks, which are already getting updated by `TaskEnd`.  I suppose its not a problem for them to get updated again with the same info, but you could probably skip the update for non-broadcast blocks, and include a comment as in `EventLoggingListener`.

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -101,25 +122,30 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
+        val (id, blockMap) = getIdAndBlockMap(rddOrBroadcastBlockId)
+        var removed: Option[BlockStatus] = None
+        if (blockMap.contains(id)) {
+          removed = blockMap(id).remove(blockId)
```
you don't need the `var` here, you can keep the `val` like before:

```
if (blockMap.contains(id)) {
  val removed = blockMap(id).remove(blockId)
  ...
```

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -166,28 +195,69 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
+
+  /**
+   * Return the number of Broadcast blocks stored in this block manager in O(RDDs) time.
+   * Note that this is much faster than `this.rddBlocks.size`, which is O(RDD blocks) time.
```
RDD should be Broadcast

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -166,28 +195,69 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
+  /**
+   * Return the number of blocks that belong to the given Broadcast variable in O(1) time.
+   * Note that this is much faster than `this.rddBlocksById(rddId).size`, which is
+   * O(blocks in this RDD) time.
```
same here

squito commented on core/src/main/scala/org/apache/spark/storage/StorageUtils.scala:
```diff
@@ -224,8 +299,14 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
+          _broadcastStorageInfo.remove(broadcastId)
+        } else {
+          _broadcastStorageInfo(broadcastId) = (newMem, newDisk, newTachyon)
+        }
```
I think you could use the same `case rddOrBroadcastBlockId @ (_: BroadcastBlockId | _: RDDBlockId)` pattern here to combine both matches in this method ... its subjective whether its worth it or not, but I think it would be cleaner.

squito commented on core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala:
```diff
@@ -39,11 +40,18 @@ private[ui] class StorageTab(parent: SparkUI) extends SparkUITab(parent, "storag
+  }
+    
+
+  /** Filter broadcast info to include only those with cached partitions */
```
delete this comment

squito commented on core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala:
```diff
@@ -79,4 +87,19 @@ class StorageListener(storageStatusListener: StorageStatusListener) extends Spar
+  override def onBlockUpdate(blockUpdateEvent: SparkListenerBlockUpdate) = synchronized {
+    // only update broadcast for now, need to be modified if want to track other blocks
+    val broadcastIdOpt = blockUpdateEvent.blockId.asBroadcastId
+    if (broadcastIdOpt.isDefined) {
```
I don't think it particularly matters here, but fyi a more "scala" way of doing this is:

```
blockUpdateEvent.blockId.asBroadcastId.foreach{broadcastBlock =>
  val broadcastId = broadcastBlock.broadcastId
  ...
}
```

squito commented on core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala:
```diff
@@ -79,4 +87,19 @@ class StorageListener(storageStatusListener: StorageStatusListener) extends Spar
   }
+
+  override def onBlockUpdate(blockUpdateEvent: SparkListenerBlockUpdate) = synchronized {
+    // only update broadcast for now, need to be modified if want to track other blocks
```
this comment is slightly misleading -- as in other places, I would instead point out that RDD blocks are handled in `onStageCompleted`, but for now we are ignoring all other blocks.

squito commented:
Hi @CodingCat , I took another look through everything.  My comments are all very minor.  But I think I'd still like to get some more thoughts from others on how the broadcast block info is passed along.  As I mentioned earlier, it is also possible to get it into `TaskEnd` events, like the RDD blocks.  I'm torn, so I'd like to hear what others think.  Sorry it is taking so long to get feedback ...

squito commented on core/src/main/scala/org/apache/spark/ui/storage/BroadcastPage.scala:
```diff
@@ -0,0 +1,86 @@
+
+private[ui] class BroadcastPage(parent: StorageTab) extends StorageDetailPage("broadcast", parent){
+
+  protected override val workerTableID: String = "broadcast-storage-by-block-table"
```
by *worker*, not by *block*

squito commented on core/src/main/scala/org/apache/spark/ui/storage/RDDPage.scala:
```diff
@@ -17,134 +17,84 @@
-    val workers = storageStatusList.map((rddId, _))
-    val workerTable = UIUtils.listingTable(workerHeader, workerRow, workers,
-      id = Some("rdd-storage-by-worker-table"))
+  protected override val workerTableID: String = "rdd-storage-by-block-table"
```
by *worker*, not by *block*

squito commented on core/src/main/scala/org/apache/spark/ui/storage/BroadcastPage.scala:
```diff
@@ -0,0 +1,86 @@
+      .map { case (blockId, status) =>
+      (blockId, status, blockLocations.get(blockId).getOrElse(Seq[String]("Unknown")))
+    }
+    (UIUtils.listingTable(blockHeader, blockRow, blocks, id = Some("rdd-storage-by-block-table")),
```
broadcast, not rdd

CodingCat committed:
address comments except merging case one

SparkQA commented:
[Test build #28577 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28577/consoleFull) for   PR 2851 at commit [`26977c9`](https://github.com/apache/spark/commit/26977c97add7d47d798985753411645e6cdcac29).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28577 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28577/consoleFull) for   PR 2851 at commit [`26977c9`](https://github.com/apache/spark/commit/26977c97add7d47d798985753411645e6cdcac29).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28577/
Test FAILed.

CodingCat committed:
display storage level of broadcast partitions

SparkQA commented:
[Test build #28585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28585/consoleFull) for   PR 2851 at commit [`49ee118`](https://github.com/apache/spark/commit/49ee1188989422673203eece83d67b590b05bb27).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28585/consoleFull) for   PR 2851 at commit [`49ee118`](https://github.com/apache/spark/commit/49ee1188989422673203eece83d67b590b05bb27).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28585/
Test PASSed.

CodingCat committed:
revise the stuffs

SparkQA commented:
[Test build #28601 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28601/consoleFull) for   PR 2851 at commit [`9b1bccf`](https://github.com/apache/spark/commit/9b1bccf89c59e95ba7132034149b0d0e657b35b7).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28601 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28601/consoleFull) for   PR 2851 at commit [`9b1bccf`](https://github.com/apache/spark/commit/9b1bccf89c59e95ba7132034149b0d0e657b35b7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28601/
Test PASSed.

CodingCat commented:
Hi, @squito , thanks for the comments, I made some revision on the patch

squito commented:
thanks for the updates!  just one small style comment.
FWIW, after thinking about it more, I am in favor of your approach of sending the block updates, just want to get some confirmation.

lgtm (with the minor style update)

CodingCat committed:
stylistic fix

SparkQA commented:
[Test build #28626 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28626/consoleFull) for   PR 2851 at commit [`1aec3a8`](https://github.com/apache/spark/commit/1aec3a8073a241d16953a0adad7a17cf3ac2bba4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28626 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28626/consoleFull) for   PR 2851 at commit [`1aec3a8`](https://github.com/apache/spark/commit/1aec3a8073a241d16953a0adad7a17cf3ac2bba4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28626/
Test PASSed.

CodingCat commented:
thanks @squito 

I updated that again

JoshRosen commented:
Hey @CodingCat, it looks like this PR is significantly out of date. Do you mind closing it for now? You can always re-open / re-submit once we're ready to work on this feature again. Thanks!

CodingCat commented:
sure, I will close this temporarily
